## First we explore the spray.csv

In [1]:
import pandas as pd
spray = pd.read_csv('../datasets/spray.csv')

In [2]:
spray.head()

Date        Time   Latitude  Longitude
0  2011-08-29  6:56:58 PM  42.391623 -88.089163
1  2011-08-29  6:57:08 PM  42.391348 -88.089163
2  2011-08-29  6:57:18 PM  42.391022 -88.089157
3  2011-08-29  6:57:28 PM  42.390637 -88.089158
4  2011-08-29  6:57:38 PM  42.390410 -88.088858

In [8]:
def explore(df):
    print(df.shape)
    print("")
    print(df.head())
    print("")
    print(df.dtypes)
    print("")
    print(df.isnull().sum())
    print("")
    print(df.describe())
    
explore(spray)

(14835, 4)

        Date     Time   Latitude  Longitude
0 2011-08-29 06:56:58  42.391623 -88.089163
1 2011-08-29 06:57:08  42.391348 -88.089163
2 2011-08-29 06:57:18  42.391022 -88.089157
3 2011-08-29 06:57:28  42.390637 -88.089158
4 2011-08-29 06:57:38  42.390410 -88.088858

Date          datetime64[ns]
Time         timedelta64[ns]
Latitude             float64
Longitude            float64
dtype: object

Date           0
Time         584
Latitude       0
Longitude      0
dtype: int64

                         Time      Latitude     Longitude
count                   14251  14835.000000  14835.000000
mean   0 days 09:12:42.373938     41.904828    -87.736690
std    0 days 00:59:54.646393      0.104381      0.067292
min           0 days 06:56:58     41.713925    -88.096468
25%    0 days 08:29:01.500000     41.785001    -87.794225
50%           0 days 09:05:28     41.940075    -87.727853
75%           0 days 09:47:33     41.980978    -87.694108
max           0 days 12:24:49     42.395983   

In [4]:
#checking the list of date that spraying occured
spray['Date'] = pd.to_datetime(spray['Date'])
spray['Time'] = pd.to_timedelta(spray['Time'])
spray.groupby(['Date']).count()

In [12]:
#checking which are the dates that time is null
nans = lambda spray: spray[spray.isnull().any(axis=1)]
nans(spray)

Date Time   Latitude  Longitude
1030 2011-09-07  NaT  41.987092 -87.794286
1031 2011-09-07  NaT  41.987620 -87.794382
1032 2011-09-07  NaT  41.988004 -87.794574
1033 2011-09-07  NaT  41.988292 -87.795486
1034 2011-09-07  NaT  41.988100 -87.796014
1035 2011-09-07  NaT  41.986372 -87.794862
1036 2011-09-07  NaT  41.986228 -87.795582
1037 2011-09-07  NaT  41.984836 -87.793998
1038 2011-09-07  NaT  41.984836 -87.794670
1039 2011-09-07  NaT  41.984884 -87.795198
1040 2011-09-07  NaT  41.982820 -87.793182
1041 2011-09-07  NaT  41.982772 -87.793950
1042 2011-09-07  NaT  41.982772 -87.795006
1043 2011-09-07  NaT  41.983444 -87.795246
1044 2011-09-07  NaT  41.984068 -87.795438
1045 2011-09-07  NaT  41.984596 -87.795534
1046 2011-09-07  NaT  41.984980 -87.795630
1047 2011-09-07  NaT  41.986708 -87.796110
1048 2011-09-07  NaT  41.987524 -87.796302
1049 2011-09-07  NaT  41.988820 -87.796638
1050 2011-09-07  NaT  41.989444 -87.796830
1051 2011-09-07  NaT  41.990068 -87.796974
1052 2011-09-07  NaT  41.988244 -87.797022
1053 2011-09-07  NaT  41.988196 -87.797742
1054 2011-09-07  NaT  41.988196 -87.798366
1055 2011-09-07  NaT  41.988196 -87.798942
1056 2011-09-07  NaT  41.988196 -87.799422
1057 2011-09-07  NaT  41.986372 -87.796494
1058 2011-09-07  NaT  41.986372 -87.797310
1059 2011-09-07  NaT  41.986372 -87.798030
...         ...  ...        ...        ...
1584 2011-09-07  NaT  41.992372 -87.816367
1585 2011-09-07  NaT  41.992468 -87.815839
1586 2011-09-07  NaT  41.992516 -87.815311
1587 2011-09-07  NaT  41.992564 -87.814735
1588 2011-09-07  NaT  41.992612 -87.814303
1589 2011-09-07  NaT  41.992612 -87.813775
1590 2011-09-07  NaT  41.992612 -87.813343
1591 2011-09-07  NaT  41.992660 -87.812767
1592 2011-09-07  NaT  41.992660 -87.812191
1593 2011-09-07  NaT  41.992612 -87.811663
1594 2011-09-07  NaT  41.992660 -87.811279
1595 2011-09-07  NaT  41.992660 -87.810847
1596 2011-09-07  NaT  41.992612 -87.810415
1597 2011-09-07  NaT  41.992660 -87.809455
1598 2011-09-07  NaT  41.992660 -87.809983
1599 2011-09-07  NaT  41.992708 -87.808927
1600 2011-09-07  NaT  41.992708 -87.808399
1601 2011-09-07  NaT  41.992756 -87.807775
1602 2011-09-07  NaT  41.992756 -87.807199
1603 2011-09-07  NaT  41.996164 -87.813727
1604 2011-09-07  NaT  41.995732 -87.813679
1605 2011-09-07  NaT  41.994916 -87.813727
1606 2011-09-07  NaT  41.992948 -87.813775
1607 2011-09-07  NaT  41.996644 -87.811567
1608 2011-09-07  NaT  41.996260 -87.811567
1609 2011-09-07  NaT  41.995876 -87.811615
1610 2011-09-07  NaT  41.995972 -87.810271
1611 2011-09-07  NaT  41.995684 -87.810319
1612 2011-09-07  NaT  41.994724 -87.810415
1613 2011-09-07  NaT  41.993092 -87.810415

[584 rows x 4 columns]

Spray conclusion: <br>
    spray.shape = (14835, 4)<br>
    584 null values in Time column only, all null values are in 2011-09-07<br>
    However, not all 2011-09-07 time are null values (584/1530 are nulls)<br>
    
    changed Date and Time to datetime and timedelta respectively<br>
    
    2 sprays in 2011, once in Aug, once in Sept<br>
    8 sprays in 2013, twice in July, 5 times in Aug, once in Sept<br>
    
sidenote: might be linked to weather conditions<br> 
(ie hotter in 2013 than 2011, or increased mosquitoes activites hence an increase in sprays)

In [624]:
spray_2 = spray

In [626]:
spray_2.to_csv("../datasets/spray_2.csv", index=False) 


## Now, we move on to explore "weather.csv"

In [440]:
weather = pd.read_csv('../datasets/weather.csv')

In [441]:
weather.head()

Station        Date  Tmax  Tmin Tavg Depart  DewPoint WetBulb Heat Cool  \
0        1  2007-05-01    83    50   67     14        51      56    0    2   
1        2  2007-05-01    84    52   68      M        51      57    0    3   
2        1  2007-05-02    59    42   51     -3        42      47   14    0   
3        2  2007-05-02    60    43   52      M        42      47   13    0   
4        1  2007-05-03    66    46   56      2        40      48    9    0   

  Sunrise Sunset CodeSum Depth Water1 SnowFall PrecipTotal StnPressure  \
0    0448   1849             0      M      0.0        0.00       29.10   
1       -      -             M      M        M        0.00       29.18   
2    0447   1850      BR     0      M      0.0        0.00       29.38   
3       -      -   BR HZ     M      M        M        0.00       29.44   
4    0446   1851             0      M      0.0        0.00       29.39   

  SeaLevel  ResultSpeed  ResultDir AvgSpeed  
0    29.82          1.7         27      9.2  
1    29.82          2.7         25      9.6  
2    30.09         13.0          4     13.4  
3    30.08         13.3          2     13.4  
4    30.12         11.7          7     11.9

In [442]:
explore(weather)

(2944, 22)

   Station        Date  Tmax  Tmin Tavg Depart  DewPoint WetBulb Heat Cool  \
0        1  2007-05-01    83    50   67     14        51      56    0    2   
1        2  2007-05-01    84    52   68      M        51      57    0    3   
2        1  2007-05-02    59    42   51     -3        42      47   14    0   
3        2  2007-05-02    60    43   52      M        42      47   13    0   
4        1  2007-05-03    66    46   56      2        40      48    9    0   

  Sunrise Sunset CodeSum Depth Water1 SnowFall PrecipTotal StnPressure  \
0    0448   1849             0      M      0.0        0.00       29.10   
1       -      -             M      M        M        0.00       29.18   
2    0447   1850      BR     0      M      0.0        0.00       29.38   
3       -      -   BR HZ     M      M        M        0.00       29.44   
4    0446   1851             0      M      0.0        0.00       29.39   

  SeaLevel  ResultSpeed  ResultDir AvgSpeed  
0    29.82          1.7     

In [444]:
#expand the Date column to Year,Month,Day to check on the period of data and see if there are any missing days/months
from pandas import Timestamp

weather['Date'] = pd.to_datetime(weather['Date'])
weather['Year'] = weather['Date'].dt.year
weather['Month'] = weather['Date'].dt.month
weather['Day'] = weather['Date'].dt.day

# Show three rows
weather.head(3)

Station       Date  Tmax  Tmin Tavg Depart  DewPoint WetBulb Heat Cool  \
0        1 2007-05-01    83    50   67     14        51      56    0    2   
1        2 2007-05-01    84    52   68      M        51      57    0    3   
2        1 2007-05-02    59    42   51     -3        42      47   14    0   

  Sunrise Sunset CodeSum Depth Water1 SnowFall PrecipTotal StnPressure  \
0    0448   1849             0      M      0.0        0.00       29.10   
1       -      -             M      M        M        0.00       29.18   
2    0447   1850      BR     0      M      0.0        0.00       29.38   

  SeaLevel  ResultSpeed  ResultDir AvgSpeed  Year  Month  Day  
0    29.82          1.7         27      9.2  2007      5    1  
1    29.82          2.7         25      9.6  2007      5    1  
2    30.09         13.0          4     13.4  2007      5    2

In [445]:
#Creating dummy variables for the weather CodeSum data
weather.CodeSum = weather.CodeSum.apply(str.split)
for (i,list_) in enumerate(weather.CodeSum):
    for item in list_:
        if item not in weather.columns:
            weather[item] = 0
            weather.set_value(col=item,index=i,value=1)

/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:7: FutureWarning: set_value is deprecated and will be removed in a future release. Please use .at[] or .iat[] accessors instead
  import sys


In [446]:
#replacing codes with other values
weather = weather.replace('M',0)
weather = weather.replace('-',np.nan)
weather = weather.replace(' ',np.nan)
weather = weather.replace('T',0.0005)

In [447]:
#dropping these columns since they are used for winter purposes,
#hence their values are zero and empty for the other weather station
weather = weather.drop(['Depth','Water1','SnowFall'],axis=1)

In [449]:
#To fill missing Tavg data with Average Temp.
for i in range(weather.shape[0]):
    if weather.loc[i,'Tavg'] == 'M':
        weather.loc[i,'Tavg'] = int(np.round((weather.loc[i,'Tmax'] + weather.loc[i,'Tmin'])/2))
    else:
        weather.loc[i,'Tavg'] = weather.loc[i,'Tavg']

In [450]:
columns = ['Tavg',
           'Depart']

for i in columns:
    weather[i] = weather[i].astype(str).astype(float)

#source: [https://www.theweatherprediction.com/habyhints/170/] <br>
A quick technique that many forecasters use to determine the wet-bulb temperature <br>
is called the "1/3 rule". The technique is to first find the dewpoint depression (temperature minus dewpoint). <br>
Then take this number and divide by 3. Subtract this number from the temperature.<br>

Since there are only 4 missing wetbulb cells, we will use this estimate<br>
wet bulb might be linked to humidity, we might be able to create a feature engineer a new column<br>

In [451]:
#To estimate the value for 'WetBulb'
for i in range(weather.shape[0]):
    if weather.loc[i,'WetBulb'] == 0:
        weather.loc[i,'WetBulb'] = np.round(weather.loc[i,'Tavg'] - (weather.loc[i,'Tavg'] - weather.loc[i,'Depart'])/3)
                                                                    
    else:
        weather.loc[i,'WetBulb'] = weather.loc[i,'WetBulb']

#Heat
A heating degree day (HDD) is a measurement designed to quantify the demand for energy 
needed to heat a building. It is the number of degrees that a day's average temperature 
is below 65o Fahrenheit (18o Celsius), which is the temperature below which buildings 
need to be heated. 

#Cool
A cooling degree day (CDD) is a measurement designed to quantify the demand for energy 
needed to cool buildings. It is the number of degrees that a day's average temperature 
is above 65o Fahrenheit (18o Celsius).

In [452]:
#To estimate the value for 'Heat' and 'Cool'
for i in range(weather.shape[0]):
    if weather.loc[i,'Heat'] == 0:
        if 65 - weather.loc[i,'Tavg'] > 0:
            weather.loc[i,'Heat'] = 65 - weather.loc[i,'Tavg']
        else:
            weather.loc[i,'Heat'] = 0 
    else:
        weather.loc[i,'Heat'] = weather.loc[i,'Heat']

        

for i in range(weather.shape[0]):
    if weather.loc[i,'Cool'] == 0:
        if  weather.loc[i,'Tavg'] - 65 > 0:
            weather.loc[i,'Cool'] = weather.loc[i,'Tavg'] - 65
        else:
            weather.loc[i,'Cool'] = 0 
    else:
        weather.loc[i,'Cool'] = weather.loc[i,'Cool']

In [453]:
#changing traces of precipation total to 0.005 so that we can use the values to analyse the model
weather['PrecipTotal'] = weather['PrecipTotal'].replace(np.nan,'0.00')
weather['PrecipTotal'] = weather['PrecipTotal'].replace('  T','0.005')

In [454]:
#change station pressure np.nan to mean
weather['StnPressure'] = weather['StnPressure'].replace(np.nan,0)
weather['StnPressure'] = weather['StnPressure'].astype(str).astype(float)
weather['StnPressure'] = weather['StnPressure'].replace(0,weather['StnPressure'].mean())

In [455]:
#change sealevel np.nan to mean
weather['SeaLevel'] = weather['SeaLevel'].replace(np.nan,0)
weather['SeaLevel'] = weather['SeaLevel'].astype(str).astype(float)
weather['SeaLevel'] = weather['SeaLevel'].replace(0,weather['StnPressure'].mean())

In [459]:
columns = ['WetBulb',
           'Heat',
           'Cool',
           'PrecipTotal',
           'AvgSpeed',
           'Sunrise',
           'Sunset']

for i in columns:
    weather[i] = weather[i].astype(str).astype(float)

In [460]:
explore(weather)

(2944, 38)

   Station       Date  Tmax  Tmin  Tavg  Depart  DewPoint  WetBulb  Heat  \
0        1 2007-05-01    83    50  67.0    14.0        51     56.0   0.0   
1        2 2007-05-01    84    52  68.0     0.0        51     57.0   0.0   
2        1 2007-05-02    59    42  51.0    -3.0        42     47.0  14.0   
3        2 2007-05-02    60    43  52.0     0.0        42     47.0  13.0   
4        1 2007-05-03    66    46  56.0     2.0        40     48.0   9.0   

   Cool  Sunrise  Sunset   CodeSum  PrecipTotal  StnPressure  SeaLevel  \
0   2.0    448.0  1849.0        []          0.0        29.10     29.82   
1   3.0      NaN     NaN        []          0.0        29.18     29.82   
2   0.0    447.0  1850.0      [BR]          0.0        29.38     30.09   
3   0.0      NaN     NaN  [BR, HZ]          0.0        29.44     30.08   
4   0.0    446.0  1851.0        []          0.0        29.39     30.12   

   ResultSpeed  ResultDir  AvgSpeed  Year  Month  Day  BR  HZ  RA  TSRA  VCTS  \
0    

In [436]:
weather.shape

(2944, 38)

In [599]:
weather_2 = weather.groupby(['Date']).mean()

In [600]:
explore(weather_2)

(1472, 36)

            Station  Tmax  Tmin  Tavg  Depart  DewPoint  WetBulb  Heat  Cool  \
Date                                                                           
2007-05-01      1.5  83.5  51.0  67.5     7.0      51.0     56.5   0.0   2.5   
2007-05-02      1.5  59.5  42.5  51.5    -1.5      42.0     47.0  13.5   0.0   
2007-05-03      1.5  66.5  47.0  57.0     1.0      40.0     49.0   8.0   0.0   
2007-05-04      1.5  72.0  50.0  29.0     2.0      41.5     50.0  36.0   0.0   
2007-05-05      1.5  66.0  53.5  60.0     2.5      38.5     49.5   5.0   0.0   

            Sunrise  Sunset  PrecipTotal  StnPressure  SeaLevel  ResultSpeed  \
Date                                                                           
2007-05-01    448.0  1849.0       0.0000       29.140    29.820         2.20   
2007-05-02    447.0  1850.0       0.0000       29.410    30.085        13.15   
2007-05-03    446.0  1851.0       0.0000       29.425    30.120        12.30   
2007-05-04    444.0  1852.0

In [602]:
columns = ['Year',
           'Month',
           'Day']

for i in columns:
    weather_2[i] = weather_2[i].astype(float).astype(int)

In [603]:
weather_2.iloc[:,20:] = weather_2.iloc[:,20:].apply(np.ceil)
weather_2.iloc[:,20:] = weather_2.iloc[:,20:].astype(float).astype(int)

In [605]:
columns = ['Sunrise',
           'Sunset']

for i in columns:
    weather_2[i] = weather_2[i].astype(float).astype(int)
    weather_2[i] = weather_2[i].astype(int).astype(str)

In [607]:
#changing Sunrise and Sunset to type - time
weather_2['Sunrise'] = str(0) + weather_2['Sunrise']
weather_2['Sunrise'] = weather_2['Sunrise'].str[0:2] + ":" + weather_2['Sunrise'].str[2:4] + ":00"

weather_2['Sunset'] = weather_2['Sunset'].str[0:2] + ":" + weather_2['Sunset'].str[2:4] + ":00"


weather_2['Sunrise'] = pd.to_timedelta(weather_2['Sunrise'])
weather_2['Sunset'] = pd.to_timedelta(weather_2['Sunset'])

In [611]:
#dropping Station column since we merge the data together
weather_2 = weather_2.drop(['Station'], axis=1)

In [616]:
weather_2['Date'] = weather_2.index

In [618]:
weather_2.to_csv("../datasets/weather_2.csv", index=False) 

weather conclusion: <br>
weather.shape = (1472, 36) <br>

change the CodeSum to dummies

# Next, to explore train and test sets

In [ ]:
train = pd.read_csv('predict-west-nile-virus/train.csv')

In [ ]:
train.head()

In [ ]:
test = pd.read_csv('predict-west-nile-virus/test.csv')

In [ ]:
test.head()

In [ ]:
sample = pd.read_csv('predict-west-nile-virus/samplesubmission.csv')

In [ ]:
sample.head()